tests/test_helpers.py

```diff
@@ -66,3 +68,59 @@ def test_local_model_non_default_name(self, tmp_path):
         # with adapter name
         result = check_if_peft_model(tmp_path / "peft-gpt2-other" / "other")
         assert result is True
+
+    def test_set_adapter_scale(self):
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        # Get the PeftModel
+        model = AutoPeftModelForCausalLM.from_pretrained("ybelkada/opt-350m-lora")
+        tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
+
+        # Prepare inputs
+        input_prompt = "Preheat the oven to 350 degrees"
+        inputs = tokenizer(input_prompt, return_tensors="pt")
+
+        # Get logits before scaling
+        with torch.no_grad():
+            inputs = inputs.to(device)
+            model = model.to(device)
+            logits_before_scaling = model(**inputs).logits
+
+        # Store the scale of the lora layers
+        scales_before_scaling = scale_from_modules(model)
+
+        with set_adapter_scale(model=model, alpha=0.5):
+            # Check for scaling
+            scales_during_scaling = scale_from_modules(model)
+            for key in scales_before_scaling.keys():
+                assert scales_before_scaling[key] != scales_during_scaling[key]
+
+            # Generate logits during scaling
+            with torch.no_grad():
+                inputs = inputs.to(device)
+                model = model.to(device)
+                logits_during_scaling = model(**inputs).logits
+
+            assert torch.allclose(logits_before_scaling, logits_during_scaling)
+
+        # Check for restored sclaes
+        scales_after_scaling = scale_from_modules(model)
+        for key in scales_before_scaling.keys():
+            assert scales_before_scaling[key] == scales_after_scaling[key]
+
+        # Get logits after scaling
+        with torch.no_grad():
+            inputs = inputs.to(device)
+            model = model.to(device)
+            logits_after_scaling = model(**inputs).logits
+
+        assert torch.allclose(logits_before_scaling, logits_after_scaling)
```

Comment:
I think the function should already work with transformers and diffusers models as is. But let's add tests for those to ensure it's really working.

Also, we need tests for:

* merging inside the context
* using multiple adapters (as the scalings of all adapters are updated)
* a test that checks that restoration works even if we have different `r` and `alpha` on the layers (using `rank_pattern` and `alpha_pattern`).

Reply:
Hey @BenjaminBossan I had some queries:
1. I see that the diffusion models with LoRAs in the Hub have a bigger base model. This in turn would be very slow for the testing pipeline. Do you have any suggestions to ease this?
2. Merging: As I understand, we merge the weights, then pass the model into the context manager, and get the merged lora outputs. After exiting from the context manager, the model should revive its merged weights. Am I understanding this correctly?
3. I am not sure I understand the usage of multiple adapters. If you could point me to some resource on this topic that would be very helpful.
4. I am unable to find an example where there is a use of `rank_pattern` and `alpha_pattern` -- Could you help me with this as well?

Reply:
1. Indeed, SD models are quite big. Right now, we still use them in tests, see here: https://github.com/huggingface/peft/blob/7e7b55880e2e0f3c80309d4e578ee36f6ff24793/tests/test_stablediffusion.py#L28.
2. No, my idea was a bit different. The test should be like this: Create a LoRA model with an adapter, set the scale to value x and get the outputs. Then, set the scale again to x and merge the weights. After exiting the context, the weights are still merged. Now when we get the output, it should be the same as our initial output. This way, we can ensure that the temporarily set scalings are correctly accounted for when merging. Does that make sense?
3. For methods like LoRA, it's possible to add multiple adapters to the same model by calling `model.add_adapter(config2)` and activate it using `set_adapter`. There should be a test to ensure that within the scaling context, the scalings of all adapters is adjusted, and that after exiting the context, all adapter scalings are correctly reset.
4. Indeed, it looks like there is a gap in our test suite when it comes to `rank_pattern` and `alpha_pattern`. But the gist of it is quite simple: If I have a `LoraConfig(target_modules="all-linear", r=8, rank_pattern={"linear_xy": 16})`, then all linear layers will have rank 8, except for "linear_xy", which will have rank 16. Same idea for `alpha_pattern`. The test should ensure that the scaling context works for these models as well.

Reply:
Thanks for the detailed reply @BenjaminBossan!

I was able to get all of the tests to work except the **merging** test. The logits after `merge_and_unload()` are not the same it seems.

Do you think I am making some mistake?

Reply:
Merging will always result in a bit of divergence because of floating point imprecision. If the logits are identical up to a precision of `atol, rtol = 1e-4, 1e-4`, it's good enough. Could you please check that?

Reply:
That was it! Made the changes in the current commit -- every test pass now :partying_face: